frontend/src/components/ProductRecommendations.jsx

```diff
@@ -17,20 +20,80 @@ function ProductRecommendations({ onProductSelect, onBack, userProfile }) {
   useEffect(() => {
     testAPIConnectivity()
     loadRecommendations()
+    loadPopularTags()
 
     // Track page view for recommendations
     if (isAuthenticated.current) {
       trackInteraction('view', null, { page: 'recommendations' });
     }
   }, [])
 
-  async function testAPIConnectivity() {
+  async function loadPopularTags() {
+    setLoadingTags(true)
     try {
-      const response = await fetch(`${process.env.REACT_APP_API_URL || 'https://skincare-api.herokuapp.com'}/products?limit=1`)
+      const tags = await SkincareAPI.getPopularTags(15) // Get top 15 tags
+      setPopularTags(tags)
+    } catch (err) {
+    } finally {
+      setLoadingTags(false)
+    }
+  }
 
-      if (response.ok) {
-        await response.json()
+  async function handleTagSelect(tag) {
+    setLoading(true)
+    setError('')
+
+    try {
+      if (selectedTags.includes(tag)) {
+        setSelectedTags(prev => prev.filter(t => t !== tag))
+
+        // If no tags left selected, load all recommendations
+        if (selectedTags.length <= 1) {
+          await loadRecommendations()
+          return
+        }
+
+        // Otherwise filter by remaining tags
+        const newSelectedTags = selectedTags.filter(t => t !== tag)
+        const filteredProducts = await filterProductsByTags(newSelectedTags)
+        setProducts(filteredProducts)
       }
+      // Otherwise add the tag to selected tags
+      else {
+        setSelectedTags(prev => [...prev, tag])
+        const filteredProducts = await filterProductsByTags([...selectedTags, tag])
+        setProducts(filteredProducts)
+      }
+    } catch (err) {
+      setError('Failed to filter products by tag. Please try again.')
+    } finally {
+      setLoading(false)
+    }
+  }
+
+  // Filter products by selected tags
+  async function filterProductsByTags(tags) {
+    if (!tags || tags.length === 0) {
+      return await SkincareAPI.getRecommendations()
+    }
+
+    // Get all products first
+    const allProducts = await SkincareAPI.getRecommendations()
+
+    // Filter products that have ALL selected tags
+    return allProducts.filter(product =>
+      tags.every(tag =>
+        product.product_tags &&
+        Array.isArray(product.product_tags) &&
+        product.product_tags.some(t => t.toLowerCase() === tag.toLowerCase())
+      )
+    )
+  }
+
+  async function testAPIConnectivity() {
+    try {
+      const response = await fetch(`/products?limit=1`);
+      await response.json();
```

Comment:
what are we doing if the APIConnectivity is not good? We can add some error handling here to denote the service is down!